Error compiling with HIP 6.3.1

When I try to compile MGARD with HIP 6.3.1, I get the following compile error:

```
[ 90%] Building HIP object CMakeFiles/mgard-library.dir/src/mgard-x/CompressionLowLevel/Compressor_1D_double_HIP.cpp.o
/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/llvm-amdgpu-6.3.1-abygvk6ih6uhr77njljb6elal4exbg7r/bin/clang++  -DMGARD_ENABLE_HIP -DMGARD_ENABLE_SERIAL -DMGARD_ZSTD -D__HIP_ROCclr__=1 -Dmgard_library_EXPORTS -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/hipcub-6.3.1-erutma6rhiuuwt2lew7wm4wsttiru2sw/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/rocprim-6.3.1-y2kptu4nrrbo3q2q5u5apysgftb5ghwk/include -I/spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include -I/spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-build-nvx7kj4/include -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/protobuf-3.28.2-4zcac5ynmrs2hbyg4oa4jxcb5frq6u4k/include -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/abseil-cpp-20240722.0-gokmudz4kba47xhtrrpesrjgxvkak5dh/include -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/zlib-ng-2.2.3-6mvcy2vb3h4g7i7ltnagzhxyrrwi6ggv/include -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/zstd-1.5.6-io4ttf7xysxy4uswvv7wk6ogupvzirhv/include -w -O3 -DNDEBUG -std=gnu++17 --offload-arch=gfx90a -fPIC -DPROTOBUF_USE_DLLS -Wnon-virtual-dtor -DNOMINMAX -o CMakeFiles/mgard-library.dir/src/mgard-x/CompressionLowLevel/Compressor_1D_double_HIP.cpp.o -x hip -c /spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-build-nvx7kj4/src/mgard-x/CompressionLowLevel/Compressor_1D_double_HIP.cpp
In file included from /spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-build-nvx7kj4/src/mgard-x/CompressionLowLevel/Compressor_1D_double_HIP.cpp:8:
In file included from /spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include/mgard-x/CompressionLowLevel/Compressor.hpp:17:
In file included from /spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include/mgard-x/CompressionLowLevel/../Hierarchy/Hierarchy.h:9:
In file included from /spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include/mgard-x/CompressionLowLevel/../Hierarchy/../RuntimeX/RuntimeX.h:37:
In file included from /spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include/mgard-x/CompressionLowLevel/../Hierarchy/../RuntimeX/DeviceAdapters/DeviceAdapterHip.h:11:
In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/hipcub-6.3.1-erutma6rhiuuwt2lew7wm4wsttiru2sw/include/hipcub/hipcub.hpp:31:
In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/hipcub-6.3.1-erutma6rhiuuwt2lew7wm4wsttiru2sw/include/hipcub/config.hpp:47:
In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/rocprim-6.3.1-y2kptu4nrrbo3q2q5u5apysgftb5ghwk/include/rocprim/device/config_types.hpp:32:
In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/rocprim-6.3.1-y2kptu4nrrbo3q2q5u5apysgftb5ghwk/include/rocprim/device/../detail/various.hpp:28:
/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/rocprim-6.3.1-y2kptu4nrrbo3q2q5u5apysgftb5ghwk/include/rocprim/device/../detail/../type_traits.hpp:372:1: error: no type named 'type' in 'rocprim::invoke_result<mgard_x::AbsMaxOp, int, double>'
  372 | using invoke_result_t = typename invoke_result<F, Args...>::type;
      | ^~~~~
/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/hipcub-6.3.1-erutma6rhiuuwt2lew7wm4wsttiru2sw/include/hipcub/backend/rocprim/block/../block/../thread/thread_operators.hpp:279:1: note: in instantiation of template type alias 'invoke_result_t' requested here
  279 | using invoke_result_t = ::rocprim::invoke_result_t<Invokable, Args...>;
      | ^
/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/hipcub-6.3.1-erutma6rhiuuwt2lew7wm4wsttiru2sw/include/hipcub/backend/rocprim/block/../block/../thread/thread_operators.hpp:283:1: note: in instantiation of template type alias 'invoke_result_t' requested here
  283 | using accumulator_t = std::decay_t<invoke_result_t<Invokable, InitT, InputT>>;
      | ^
/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/hipcub-6.3.1-erutma6rhiuuwt2lew7wm4wsttiru2sw/include/hipcub/backend/rocprim/block/../block/../thread/thread_operators.hpp:357:25: note: in instantiation of template type alias 'accumulator_t' requested here
  357 |     using accum_type  = accumulator_t<BinaryFunction, init_type, input_type>;
      |                         ^
/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/hipcub-6.3.1-erutma6rhiuuwt2lew7wm4wsttiru2sw/include/hipcub/backend/rocprim/device/device_reduce.hpp:188:13: note: in instantiation of template class 'hipcub::detail::convert_binary_result_type_wrapper<int, double *, double *, mgard_x::AbsMaxOp>' requested here
  188 |             ::hipcub::detail::convert_binary_result_type<T, InputIteratorT, OutputIteratorT>(
      |             ^
/spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include/mgard-x/CompressionLowLevel/../Hierarchy/../RuntimeX/DeviceAdapters/DeviceAdapterHip.h:2333:27: note: in instantiation of function template specialization 'hipcub::DeviceReduce::Reduce<double *, double *, mgard_x::AbsMaxOp, int, unsigned long>' requested here
 2333 |     hipcub::DeviceReduce::Reduce(d_temp_storage, temp_storage_bytes, v.data(),
      |                           ^
/spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include/mgard-x/CompressionLowLevel/NormCalculator.hpp:44:35: note: in instantiation of function template specialization 'mgard_x::DeviceCollective<mgard_x::HIP>::AbsMax<double>' requested here
   44 |     DeviceCollective<DeviceType>::AbsMax(
      |                                   ^
/spack-build/root/spack-stage-mgard-2023-12-09-nvx7kj4mmohpueswosincok5gcklcc67/spack-src/include/mgard-x/CompressionLowLevel/Compressor.hpp:106:9: note: in instantiation of function template specialization 'mgard_x::norm_calculator<(unsigned char)'\x01', double, mgard_x::HIP>' requested here
  106 |         norm_calculator(original_data, SubArray(norm_tmp_array),
      |         ^
1 error generated when compiling for gfx90a.
```

The problem seems to be that the arguments to the `invoke_result_t` template not matching correctly.

--- a/include/mgard-x/RuntimeX/DeviceAdapters/DeviceAdapterHip.h
+++ b/include/mgard-x/RuntimeX/DeviceAdapters/DeviceAdapterHip.h
@@ -2341,7 +2341,8 @@ public:
     AbsMaxOp absMaxOp;
     hipStream_t stream = DeviceRuntime<HIP>::GetQueue(queue_idx);
     hipcub::DeviceReduce::Reduce(d_temp_storage, temp_storage_bytes, v.data(),
-                                 result.data(), n, absMaxOp, 0, stream);
+                                 result.data(), n, absMaxOp, static_cast<T>(0),
+                                 stream);
     ErrorAsyncCheck(hipGetLastError(), "DeviceCollective<HIP>::AbsMax");
     if (DeviceRuntime<HIP>::SyncAllKernelsAndCheckErrors) {
       ErrorSyncCheck(hipDeviceSynchronize(), "DeviceCollective<HIP>::AbsMax");

